src/libraries/System.Private.CoreLib/src/System/Math.DivModInt.cs

```diff
@@ -0,0 +1,210 @@
+        {
+            if (Is32BitSigned(divisor))
+            {
+                if ((int)divisor == 0)
+                {
+                    ThrowHelper.ThrowDivideByZeroException();
+                }
+
+                if ((int)divisor == -1)
+                {
+                    if ((ulong)dividend == 0x8000000000000000ul)
+                    {
+                        ThrowHelper.ThrowOverflowException();
+                    }
+                    return -dividend;
+                }
+
+                // Check for -ive or +ive numbers in the range -2**31 to 2**31
+                if (Is32BitSigned(dividend))
+                {
+                    return (int)dividend / (int)divisor;
+                }
+            }
+
+            return InternalDivInt64(dividend, divisor);
+        }
+
+        [StackTraceHidden]
+        internal static ulong DivUInt64(ulong dividend, ulong divisor)
+        {
+            if (Hi32Bits(divisor) == 0)
+            {
+                if ((uint)divisor == 0)
+                {
+                    ThrowHelper.ThrowDivideByZeroException();
+                }
+
+                if (Hi32Bits(dividend) == 0)
+                    return (uint)dividend / (uint)divisor;
+            }
+
+            return InternalDivUInt64(dividend, divisor);
+        }
+
+        [StackTraceHidden]
+        internal static int ModInt32(int dividend, int divisor)
+        {
+            if ((uint)(divisor + 1) <= 1) // Unsigned test for divisor in [-1 .. 0]
+            {
+                if (divisor == 0)
+                {
+                    ThrowHelper.ThrowDivideByZeroException();
+                }
+                else if (divisor == -1)
+                {
+                    if (dividend == int.MinValue)
+                    {
+                        ThrowHelper.ThrowOverflowException();
+                    }
+                    return 0;
+                }
+            }
+
+            return InternalModInt32(dividend, divisor);
+        }
+
+        [StackTraceHidden]
+        internal static uint ModUInt32(uint dividend, uint divisor)
+        {
+            if (divisor == 0)
+            {
+                ThrowHelper.ThrowDivideByZeroException();
+            }
+
+            return InternalModUInt32(dividend, divisor);
+        }
+
+        [StackTraceHidden]
+        internal static long ModInt64(long dividend, long divisor)
+        {
+            if (Is32BitSigned(divisor))
+            {
+                if ((int)divisor == 0)
+                {
+                    ThrowHelper.ThrowDivideByZeroException();
+                }
+
+                if ((int)divisor == -1)
+                {
+                    // <TODO>TODO, we really should remove this as it lengthens the code path
+                    // and the spec really says that it should not throw an exception. </TODO>
+                    if ((ulong)dividend == 0x8000000000000000ul)
```

Comment:
@tannergooding, I adapted this from jithelpers.cpp, is this TODO something we should address while at it? I'm assuming none of the 64-bit platforms (including riscv/loongarch) possess this behavior either.

Reply:
x64 does have this behavior:
```
class Test
{
    static long dividend = unchecked((long)0x8000000000000000);
    static long divisor = -1;

    static void Main() {
        Console.WriteLine(dividend % divisor);
    }
}
```
throws System.OverflowException on x64. I have not checked other archs, but I expect it to be the same.

Reply:
Ah, thanks. I think we can then just delete this TODO?

Reply:
I think so